zetaclient/chains/solana/signer/signer.go

```diff
@@ -0,0 +1,280 @@
+	"github.com/zeta-chain/zetacore/zetaclient/outboundprocessor"
+)
+
+var _ interfaces.ChainSigner = (*Signer)(nil)
+
+// Signer deals with signing BTC transactions and implements the ChainSigner interface
+type Signer struct {
+	*base.Signer
+
+	// client is the Solana RPC client that interacts with the Solana chain
+	client interfaces.SolanaRPCClient
+
+	// solanaFeePayerKey is the private key of the fee payer account on Solana chain
+	solanaFeePayerKey solana.PrivateKey
+
+	// gatewayID is the program ID of gateway program on Solana chain
+	gatewayID solana.PublicKey
+
+	// pda is the program derived address of the gateway program
+	pda solana.PublicKey
+}
+
+// NewSigner creates a new Bitcoin signer
+func NewSigner(
+	chain chains.Chain,
+	chainParams observertypes.ChainParams,
+	solClient interfaces.SolanaRPCClient,
+	tss interfaces.TSSSigner,
+	solanaKey solana.PrivateKey,
+	ts *metrics.TelemetryServer,
+	logger base.Logger,
+) (*Signer, error) {
+	// create base signer
+	baseSigner := base.NewSigner(chain, tss, ts, logger)
+
+	// parse gateway ID and PDA
+	gatewayID, pda, err := contract.ParseGatewayIDAndPda(chainParams.GatewayAddress)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot parse gateway address %s", chainParams.GatewayAddress)
+	}
+	logger.Std.Info().Msgf("Solana fee payer address: %s", solanaKey.PublicKey())
+
+	// create solana observer
+	return &Signer{
+		Signer:            baseSigner,
+		client:            solClient,
+		solanaFeePayerKey: solanaKey,
+		gatewayID:         gatewayID,
+		pda:               pda,
+	}, nil
+}
+
+// SignMsgWithdraw signs a message for Solana gateway withdraw/withdraw_spl instruction
+func (signer *Signer) SignMsgWithdraw(
+	ctx context.Context,
+	params *types.OutboundParams,
+	height uint64,
+) (*contract.MsgWithdraw, error) {
+	chain := signer.Chain()
+	// #nosec G115 always positive
+	chainID := uint64(signer.Chain().ChainId)
+	nonce := params.TssNonce
+	amount := params.Amount.Uint64()
+
+	// check receiver address
+	to, err := chains.DecodeSolanaWalletAddress(params.Receiver)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot decode receiver address %s", params.Receiver)
+	}
+
+	// prepare withdraw msg and compute hash
+	msg := contract.NewMsgWithdraw(chainID, nonce, amount, to)
+	msgHash := msg.Hash()
+
+	// sign the message with TSS to get an ECDSA signature.
+	// the produced signature is in the [R || S || V] format where V is 0 or 1.
+	signature, err := signer.TSS().Sign(ctx, msgHash[:], height, nonce, chain.ChainId, "")
+	if err != nil {
+		return nil, errors.Wrap(err, "Key-sign failed")
+	}
+	signer.Logger().Std.Info().Msgf("Key-sign succeed for chain %d nonce %d", chainID, nonce)
+
+	// attach the signature and return
+	return msg.WithSignature(signature), nil
+}
+
+// SignWithdrawTx signs the Solana gateway 'withdraw' transaction specified by 'msg'
+func (signer *Signer) SignWithdrawTx(ctx context.Context, msg contract.MsgWithdraw) (*solana.Transaction, error) {
```

Comment:
Unclear what is the difference with `SignMsgWithdraw`, we should explain better in comments and potentially consider some renaming

Reply:
I slightly modified the description. The `SignMsgWithdraw` is to sign a message with TSS key and returns a message object. The  `SignWithdrawTx` wraps the msg into a Solana transaction and signs with fee payer key.

Reply:
https://github.com/zeta-chain/node/pull/2560/commits/ada365437e0782d6d2f6725ca87d6c53a687118d